crates/example-types/src/block_types.rs

```diff
@@ -17,38 +17,77 @@ use hotshot_types::{
 };
 use serde::{Deserialize, Serialize};
 use sha3::{Digest, Keccak256};
+use snafu::Snafu;
 use time::OffsetDateTime;
 
 use crate::{node_types::TestTypes, state_types::TestInstanceState};
 
 /// The transaction in a [`TestBlockPayload`].
 #[derive(Default, PartialEq, Eq, Hash, Serialize, Deserialize, Clone, Debug)]
-pub struct TestTransaction(pub Vec<u8>);
+#[serde(try_from = "Vec<u8>")]
+pub struct TestTransaction(Vec<u8>);
+
+#[derive(Debug, Snafu)]
+pub struct TransactionTooLong;
+
+impl TryFrom<Vec<u8>> for TestTransaction {
+    type Error = TransactionTooLong;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_new(value).ok_or(TransactionTooLong)
+    }
+}
 
 impl TestTransaction {
+    /// Construct new transaction
+    ///
+    /// # Panics
+    /// If `bytes.len()` > `u32::MAX`
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self::try_new(bytes).expect("Vector too long")
```

Comment:
This `expect` may be controversial. My reasoning was
- 4Gb transaction is an absurd notion we shouldn't ever encounter
- Rust stdlib is allowed to panic on allocation failure, this isn't that different of a case
- this is a testing type anyway, no production code will ever call this function

Reply:
That's a reasonable justification.